### ISSUE TITLE
Multiple Line Paste/Multiple Line Text

### ISSUE THREAD

**Issue by Snuupy** [hexchat/hexchat#85]
Multiple line pastes aren't being shown as multiple lines in the textbox that can be typed in.

Whenever a line is pasted into HexChat, it looks like this: http://i.imgur.com/wGlam.png

This is a multi line paste into the HexChat textbar: http://i.imgur.com/4zvBF.png
This is what it looks like in the channel: http://i.imgur.com/7hxFQ.png

Currently, there's no way to distinguish a message that will take multiple lines, and a message that will take one line.

I'd much prefer it to show the new lines. In ChatZilla, whenever someone wants to post multiple lines, a textbox with multiple lines is shown: http://i.imgur.com/7l4Ry.png

**Comment by Vastrix** [comment]
1 year 3 months and 22 days later..
Is "soon" over yet? :(

**Comment by FichteFoll** [comment]
I would also like to raise priority for this issue, any multi-line content is a pain in the input box, even deleting it because selections become weird.

**Comment by Arnavion** [comment]
I looked at compiling gtkspell for Windows 2 weeks ago and it was too much of a pain in the ass to do, so no.

**Comment by Vastrix** [comment]
![Please :(](http://iconbug.com/data/82/256/40bac1cac7587f41a94c8f9def53549a.png)

**Comment by TingPing** [comment]
gtkspell support was already removed so that comment is no longer relevant.

**Comment by Vastrix** [comment]
So, no hope at all for this feature request? :cry:

**Comment by TingPing** [comment]
No hope.

**Comment by Vastrix** [comment]
![Ok then :(](http://1.bp.blogspot.com/-3G_Xx0uE8_U/TcG2-V_-FOI/AAAAAAAAAE4/VyMaT1o63bE/s1600/alice%2527s-third-crying-scene.gif) so be it ;(

**Comment by Poorchop** [comment]
>No hope.

I don't think that's entirely true. Keep in mind that HexChat is a free software project so anyone is free to make changes or add new features. If you're willing to figure out the hard parts, you're fully capable of adding this yourself.

**Comment by whispy** [comment]
@TingPing @Arnavion  What about [PyGTKSpellcheck](https://github.com/koehlma/pygtkspellcheck)? It's a drop-in replacement for GTKSpell that, I think, has much better support for Windows.

Better multiline support is sorely needed in Hexchat! I can't count the number of times I've accidentally spammed a channel due to invisible line breaks, or have been unable to edit typos I've made due to text going off the side of the entry field.

**Comment by Arnavion** [comment]
And how exactly do you expect a C application to use a Python library?

**Comment by whispy** [comment]
https://docs.python.org/3.1/extending/embedding.html

> It is also possible to do it the other way around: enrich your C/C++ application by embedding Python in it. Embedding provides your application with the ability to implement some of the functionality of your application in Python rather than C or C++.

While it's up to the you and the Hexchat developers to decide whether this is too much overhead or too risky, it's certainly possible to implement.

**Comment by Arnavion** [comment]
Yes thanks, I know how to embed Python. That's how we do the Python plugin after all.

That doesn't answer my question.

**Comment by TingPing** [comment]
That python library has literally no relevance to anything in this issue. HexChat is in C, it would (did) use gtkspell.

**Comment by CyberShadow** [comment]
> No hope.

Sorry, could you please describe what is the current situation and what is actually blocking this? Is it something an outside developer could reasonably solve in a day or two?

**Comment by TingPing** [comment]
>  Is it something an outside developer could reasonably solve in a day or two?

To create a solution I would be satisfied with would mean making a custom widget, so if you are very familiar with Gtk maybe you could get something going in a week. Still being gtk2 makes things a lot harder though.

**Comment by CyberShadow** [comment]
Why is a custom widget required? Spell checking? Does Gtk2 not have an equivalent of a multi-line rich text control that could be adapted for this?

**Comment by TingPing** [comment]
Because we want the best of both worlds, a widget that behaves like a normal GtkEntry but one that accepts multiple lines. GtkTextView is a "multi-line rich text control" which means its a garbage entry box. By widget I do mostly mean a subclass of one of those extending them not something from scratch.

**Comment by CyberShadow** [comment]
Sorry, why is GtkTextView not suitable? Different worlds but e.g. I know mIRC successfully uses a Windows rich text control for multi-line input and it acts just like a regular one.

**Comment by TingPing** [comment]
It sucks at dynamic sizing, I want something that starts at one line height and grows as you insert text, to a certain point ofc. I don't want a sloppy giant entry box. On top of that I would like to theme it like a normal GtkEntry (basically not possible in Gtk2).

**Comment by Fironet** [comment]
> I don't want a sloppy giant entry box.

A manually resizable multi-line text box would be perfect for me, actually :)

**Comment by hxe** [comment]
how about a little toggle button on the side of the input box which when enabled allows the input box to expand when filled with newlines? and optionally make it not interpret the input as commands, or i guess that could be another button, and solve https://github.com/hexchat/hexchat/issues/1606 . a button to toggle multiline input view and a button to toggle between regular irc input and only text (dont interpret any lines as commands)

**Comment by seema84** [comment]
4 years and the problem is not fixed? wow

**Comment by Poorchop** [comment]
I don't understand why seema's comment was pruned but in any case, if you want something done, then feel free to do it yourself. If you have a solution, then submit the patch.

**Comment by madnight** [comment]
> No hope.

![](https://i.imgur.com/EvKqWh2.png)
There is some hope ... : )

I used a GtkTextView that looks like a TextEntry and increases in size if you reach the line end.
Which is basically:
`gtk_text_view_set_wrap_mode(GTK_TEXT_VIEW (textview), GTK_WRAP_WORD_CHAR);`

Posting messages works, but im still struggling with spell checking and other features like tab name completion ...

**Comment by seebs** [comment]
I feel I should point out, I would be 100% okay with a multi-line text input which didn't support spell-checking, tab completion, or whatever. Because I don't use those, but I am really missing multi-line input. (Also relevant: Shift-up-arrow to select current text, and alt-enter/alt-return to insert a newline without sending the message yet.)

**Comment by TingPing** [comment]
As I've mentioned before it isn't a question of features it is a question of UX design. It needs to size nicely, behave as expected, not look awful with 0 padding, etc.

**Comment by madnight** [comment]
seebs, i think i can provide a version that would satisfy you (and me), but it wouldn't be enough to fullfill TingPings UX requirements, hence not offical

**Comment by SalmanPK** [comment]
I hope this can be implemented somehow 😢

**Comment by aphirst** [comment]
I was about to make a new Feature Request issue, but since I found this one, and since it's been a very long time since the last activity, figured I'd instead show my interest.

I personally find it very frustrating when I type longer sentences that it's impossible to "check" what I've written without painful manual scrolling through the single line, or copying the entire message out into an external editor. This often causes me to make grammatical errors or typos, or even lose my entire train of thought; and since I have a small screen, even expecting me to "just widen the window" isn't really a reasonable option.

Being able to have an automatically vertically-expanding input field would be for me a huge quality-of-life improvement, and it's a feature that in many other areas I simply take for granted (even pidgin has it).

I echo @seebs' position in that the multi-line input seems much more important than the other features. If it's a problem that really can't be solved yet, would it not at least be possible to have it as an option (with a disclaimer) ?

**Comment by UMLAUTaxl** [comment]
15thed

**Comment by Snuupy** [comment]
For everyone following this thread, I have personally switched to [thelounge](https://github.com/thelounge/) as it is cloud hosted, runs as a PWA, and synchronizes everything with push notifications. It does support multiple lines (shift + enter), but was not designed with native OS APIs.

If you use multiple devices like me, you may find it is not worth your while to run native clients on every machine that you own anymore (ex. 2 laptops, a phone) and may prefer to have it "in the cloud".

**Comment by LorenzoAncora** [comment]
10 years have passed since the initial feature request.
I _suggest_ that, if multi-line isn't going to be implemented, it would be wise to at least notify the user when:
- the message is too long and won't be sent as shown;
- the message contains invisible characters or newlines.

The goal of a chat client is to allow fast real-time communication, so end users should be put into the position to write with confidence. I think a simple icon at the end of the message entry box would suffice (maybe [set_icon_from_gicon](https://docs.gtk.org/gtk3/method.Entry.set_icon_from_gicon.html) ?)
